Ex4/Ex4.java

```diff
@@ -0,0 +1,81 @@
+public class Ex4 {
+
+	public static final boolean LAND = true;
+	public static final boolean WATER = false;
+	
+	public static boolean[][] map = {{WATER, LAND, WATER, LAND},
+					 {LAND, LAND, WATER, WATER},
+					 {WATER, WATER, LAND, WATER},
+					 {WATER, WATER, LAND, WATER}};
+	
+	public static void main(String[] args) {
+		System.out.print("Number of islands in the map: ");
+		System.out.println(countIslands(map));
+	}
+	
+	/**
+	 * iterate over each cell of the map.
+	 * when encountering land, mark all of its consecutive land neighbors
+	 * to form one island
+	 * @param map boolean map where true is land and false is water
+	 * @return the number of islands found
+	 */
+	public static int countIslands(boolean[][] map) {
+		int numberOfIslands = 0;
+		
+		for (int row = 0; row < map.length; row++) {
+			for (int col = 0; col < map[0].length; col++) {
+				if (map[row][col]) {
+					exploreNeighbors(row, col, map);
+					numberOfIslands++;
+				}
+			}
+		}
+		return numberOfIslands;
+	}
+
+	/**
+	 * Mark adjacent land tiles of specific cell coordinate in the map
+	 * as contiguous land to form an island and avoid counting it twice.
+	 * @param row
+	 * @param col
+	 * @param map
+	 */
```

Comment:
Nice clean implementation of flood fill.
One thing which is dangerous is you are relying on a mutable data structure (a boolean [][]) for correctness of your algorithm. In multi-threaded code, the caller could modify this array while you are using it, and you'd get a really hard to discover bug.

Or the caller might rely on you leaving the input unchanged (say they assumed they could then print the map).

It is a good practice to either take a copy of input parameters, or make them immutable (see Guava's ImmutableSet etc) for safety, and changing input parameters is always dangerous. You should very rarely do it, but if you do it should be clearly explained in the javadoc.